src/basic.jl

```diff
@@ -43,20 +48,25 @@ This should usually be an empty `Vector{S}` where `S` is the type of the state f
 """
 function particle_memory end
 
-function update(up::BasicParticleFilter, b::ParticleCollection, a, o)
+function update(up::BasicParticleFilter, b::AbstractParticleBelief, a, o)
     pm = up._particle_memory
     wm = up._weight_memory
+    if (!isempty(wm) && calculate_ess(wm) < up.resampling_threshold)
```

Comment:
Why do we need to check if `wm` is empty? Shouldn't `calculate_ess` prevent resampling on the first step? Also, the same updater might be re-used for multiple simulations, so this is not a reliable way to check.

(In general, to write maintainable code, it is not good to check implicit side effects (e.g. `wm` will probably be empty at the beginning, but that is not guaranteed) it is better to explicitly check whether it is the first step :smile: but, by design, that is not possible)

Reply:
Sorry it took me some time to figure out why I had this conditional statement. I realized I was testing this after calling the `resize!` function with an empty list (the initial case), which causes wm to have actual small weights such that `calculate_ess` is always lower than the `resampling_threshold`. 

Now `calculate_ess` is called before the `resize!` function and prevents resampling on the first step as it returns `Inf`